### PR TITLE
Refactoring tf modules and add google example

### LINKED ISSUE #75: Create google kubernetes engine terraform module

### PR DESCRIPTION
**Pull request by nhinze23**
Resolves #75

### CODE REVIEW

File: terraform/examples/ces_google_gke/README.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+Check plan
+`terraform plan -var-file=secretVars.tfvars -var-file=vars.tfvars -target=module.google_gke`
+
+Apply with
+`terraform apply -var-file=secretVars.tfvars -var-file=vars.tfvars -target=module.google_gke`
+
+This takes up to 15 minutes.
+
+Create remaining resources
```
Comment (robertauer) [inline terraform/examples/ces_google_gke/README.md:52]:
```suggestion
Create the remaining resources, including the CES setup
```

File: terraform/examples/ces_google_gke/README.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+`terraform apply -var-file=secretVars.tfvars -var-file=vars.tfvars`
+
+# Get kubeconfig
+
+- `gcloud container clusters get-credentials <cluster_name> --zone europe-west3-c --project $PROJECT_ID`
```
Comment (robertauer) [inline terraform/examples/ces_google_gke/README.md:57]: Doesn't the kubeconfig_generator module generate the kubeconfig?

File: terraform/examples/ces_google_gke/README.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+# Delete cluster
+
+- `terraform destroy -var-file=secretVars.tfvars -var-file=vars.tfvars`
+  or `terraform destroy -var-file=secretVars.tfvars -var-file=vars.tfvars`
+
+# Backup configuration
+
+## Bucket configuration
+
+### If you plan to use backup & restore with a google bucket firstly you need to create a separate service account.
```
Comment (robertauer) [inline terraform/examples/ces_google_gke/README.md:68]:
```suggestion
### If you plan to use backup & restore with a google bucket, you need to create a separate service account first.
```

File: terraform/examples/ces_google_gke/README.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+And assign the necessary Roles
+
+`gcloud projects add-iam-policy-binding $PROJECT_ID --member="serviceAccount:$SERVICE_ACCOUNT_NAME@$PROJECT_ID.iam.gserviceaccount.com" --role="roles/container.serviceAgent" --role="roles/editor"`
+
+Get that service account and save it to `secrets/gcp_sa.json`:
+
+`gcloud iam service-accounts keys create secrets/gcp_sa.json --iam-account=$SERVICE_ACCOUNT_NAME@$PROJECT_ID.iam.gserviceaccount.com`
+
+# General configuration (not mandatory)
```
Comment (robertauer) [inline terraform/examples/ces_google_gke/README.md:32]:
```suggestion
# General configuration
Use the `vars.tfvars.template` file to create `vars.tfvars` and set your GCP project and cluster name in it.
```

File: terraform/google_cloud_storage_bucket/main.tf [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  purpose = var.key_purpose
+
+  lifecycle {
+    prevent_destroy = false # TODO Change this to true
+  }
+}
+
+#data "google_storage_project_service_account" "gcs_account" {
+#
+#}
```
Comment (robertauer) [inline terraform/google_cloud_storage_bucket/main.tf:42]: Write a comment how this is maybe useful or delete it.

File: terraform/kubeconfig_generator/README.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
+# Kubeconfig Terraform module
+
+This module can generate a kubeconfig file for accessing a kubernetes cluster.
+
+## setup with GCloud
```
Comment (robertauer) [inline terraform/kubeconfig_generator/README.md:5]:
```suggestion
## Setup with GCloud
```

File: terraform/kubeconfig_generator/README.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  cluster_ca_certificate = module.google_gke.ca_certificate
+  cluster_endpoint       = "https://${module.google_gke.endpoint}"
+
+  kubeconfig_path = "kubeconfig"
+}
+```
+
+Be aware that this way a short-lived access token is used. The token will be updated at every further `terraform apply`.
+
+## setup with Azure
```
Comment (robertauer) [inline terraform/kubeconfig_generator/README.md:22]:
```suggestion
## Setup with Azure
```

File: terraform/kubeconfig_generator/README.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+Notice that azure has its own way to get the kubeconfig via `azurerm_kubernetes_cluster.aks.kube_config_raw`.
```
Comment (robertauer) [inline terraform/kubeconfig_generator/README.md:24]:
```suggestion
Notice that Azure has its own way to get the kubeconfig via `azurerm_kubernetes_cluster.aks.kube_config_raw`.
```

File: terraform/kubeconfig_generator/README.md [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+You can write it to a file with:
+```terraform
+resource "local_sensitive_file" "kubeconfig" {
+  content  = azurerm_kubernetes_cluster.aks.kube_config_raw
+  filename = "my/.kube/config"
+}
+```
+
+## get kubeconfig
```
Comment (robertauer) [inline terraform/kubeconfig_generator/README.md:33]:
```suggestion
## Get kubeconfig
```

**Review by robertauer — CHANGES_REQUESTED** [review]

File: terraform/azure_aks/README.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
+# Usage
+
+# Secret configuration
+- Get Azure credentials
+  - `az login --scope https://graph.microsoft.com//.default`
```
[the remark this answers is not in the document]
Comment (nhinze23) [reply terraform/azure_aks/README.md:5]: No this is correct.. :laughing: 
Try to create a service account without login:
- `az ad sp create-for-rbac --skip-assignment`
and this message will appear:
`Interactive authentication is needed. Please run:
az login --scope https://graph.microsoft.com//.default
`

**Review by robertauer — APPROVED** [review]

**Pull request merged**